### ISSUE TITLE
[BUG] MONAI Deploy App Stuck Due to Downstream Operator Not Receiving Data*

### ISSUE THREAD

**Issue by Omar-Faisal** [Project-MONAI/monai-deploy-app-sdk#531]
# **Issue: MONAI Deploy App Stuck Due to Downstream Operator Not Receiving Data**

## **Description**
I am working on deploying a **Bone Age Prediction Model** using **MONAI Deploy App SDK**, and I am encountering an issue where my MONAI pipeline is not passing data downstream correctly. Specifically, the application starts successfully, but execution is halted due to the **"No receiver connected to transmitter"** error. The execution graph fails to tick, leading to a **deadlock.**

## **Expected Behavior**
The application should:
1. **Load a DICOM Study** from the specified input path.
2. **Select the correct DICOM series** using the `DICOMSeriesSelectorOperator`.
3. **Convert the series into a volume** using `DICOMSeriesToVolumeOperator`.
4. **Run inference** on the extracted image using `DICOMBoneAgeOperator`.
5. **Output the predicted bone age** in months.

## **Actual Behavior**
- The pipeline initializes, and the operators are properly connected.
- However, during execution, the error message appears:
  
  ```
  [error] [entity_executor.cpp:309] [E00025] No receiver connected to transmitter of DownstreamReceptiveSchedulingTerm 30 of entity "unnamed_operator_1". The entity will never tick.
  ```

- The scheduler stops because the pipeline **does not progress past the data loading stage**:
  
  ```
  2025-03-12 08:01:17.596 INFO  gxf/std/greedy_scheduler.cpp@372: Scheduler stopped: Some entities are waiting for execution, but there are no periodic or async entities to get out of the deadlock.
  ```

- The DICOM loader outputs the correct ports, and the pipeline **seems correctly connected**, but no data flows through the pipeline.

## **Environment Details**
- **Operating System**: Ubuntu 22.04
- **Python Version**: 3.8
- **MONAI Deploy SDK Version**: Latest
- **Installed Dependencies**:
  ```bash
  pip install gdcm pylibjpeg pylibjpeg-libjpeg pylibjpeg-openjpeg pylibjpeg-rle
  ```

## **Logs**
```
🚀 Starting Bone Age Prediction App...
[info] [fragment.cpp:586] Loading extensions from configs...
[2025-03-12 08:01:17,396] [INFO] (root) - Parsed args: Namespace(argv=['app.py', '--input', '/home/omar/bone_age_deploy/monai_deploy_app/input', '--output', '/home/omar/bone_age_deploy/monai_deploy_app/output'], input=PosixPath('/home/omar/bone_age_deploy/monai_deploy_app/input'), log_level=None, model=None, output=PosixPath('/home/omar/bone_age_deploy/monai_deploy_app/output'), workdir=None)
[2025-03-12 08:01:17,397] [INFO] (BoneAgePredictionApp) - ✅ Output ports of dicom_loader: ['dicom_study_list']
[2025-03-12 08:01:17,397] [INFO] (BoneAgePredictionApp) - ✅ Output ports of series_selector: ['study_selected_series_list']
[2025-03-12 08:01:17,397] [INFO] (BoneAgePredictionApp) - ✅ Output ports of series_to_volume: ['image']
[2025-03-12 08:01:17,397] [INFO] (BoneAgePredictionApp) - ✅ Input ports of dicom_processor: ['image']
[2025-03-12 08:01:17,397] [INFO] (BoneAgePredictionApp) - ✅ Operators successfully connected.
[info] [gxf_executor.cpp:252] Creating context
[info] [gxf_executor.cpp:1974] Activating Graph...
[error] [entity_executor.cpp:309] [E00025] No receiver connected to transmitter of DownstreamReceptiveSchedulingTerm 30 of entity "unnamed_operator_1". The entity will never tick.
```

## **Full Code of `app.py`**
```python
import logging
from pathlib import Path
import sys
import os

sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))

import torch
import numpy as np
import pydicom
from monai.deploy.conditions import CountCondition
from monai.deploy.core import Application, Operator
from monai.deploy.operators import (
    DICOMDataLoaderOperator,
    DICOMSeriesSelectorOperator,
    DICOMSeriesToVolumeOperator
)
from inference.inference import BoneAgeInference

CONFIG = {
    "model_name": "microsoft/swin-large-patch4-window7-224",
    "feature_dim": 1024,
    "dropout": 0.3,
    "use_cuda": torch.cuda.is_available()
}

MODEL_WEIGHTS_PATH = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "model", "best_model.pth"))

inference_model = BoneAgeInference(MODEL_WEIGHTS_PATH, CONFIG)

class DICOMBoneAgeOperator(Operator):
    def __init__(self, fragment):
        super().__init__(fragment)
        self.logger = logging.getLogger(type(self).__name__)
    
    def setup(self, spec):
        spec.input("image")
        spec.output("predicted_age")
    
    def compute(self, input_context, output_context, execution_context):
        dicom_image = input_context.receive("image")
        if dicom_image is None:
            raise ValueError("No image received for processing.")
        pixel_array = dicom_image.pixel_array.astype(np.float32)
        image_np = np.uint8(255 * (pixel_array / np.max(pixel_array)))
        predicted_age = inference_model.predict_from_numpy(image_np, "M")
        output_context.write("predicted_age", predicted_age)

class BoneAgePredictionApp(Application):
    def compose(self):
        app_context = Application.init_app_context(self.argv)
        app_input_path = Path(app_context.input_path)
        app_output_path = Path(app_context.output_path)
        dicom_loader = DICOMDataLoaderOperator(self, CountCondition(self, 1), input_folder=app_input_path, force=True)
        series_selector = DICOMSeriesSelectorOperator(self)
        series_to_volume = DICOMSeriesToVolumeOperator(self)
        dicom_processor = DICOMBoneAgeOperator(self)
        self.add_operator(dicom_loader)
        self.add_operator(series_selector)
        self.add_operator(series_to_volume)
        self.add_operator(dicom_processor)
        self.add_flow(dicom_loader, series_selector, {("dicom_study_list", "dicom_study_list")})
        self.add_flow(series_selector, series_to_volume, {("study_selected_series_list", "study_selected_series_list")})
        self.add_flow(series_to_volume, dicom_processor, {("image", "image")})

if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO, format="%(message)s")
    app = BoneAgePredictionApp()
    app.run()
```

## **Question**
- How can I resolve the **deadlock** error?
- Could there be a **missing output definition** in one of the operators?

Any guidance would be greatly appreciated! 🚀

**Comment by MMelQin** [comment]
@Omar-Faisal sorry for the late reply. This issue will be investigated ASAP.
